### ISSUE TITLE
Logrotate causes Rsyslog to hang 

### ISSUE THREAD

**Issue by YMAHPro** [rsyslog/rsyslog#5392]
### Expected behavior
Rsyslog detecting when log file is rotated because of logrotate with copytruncate and after  statefile deletion and rsyslog restart, meaning statefile should reset to "curr_offs": 0, "strt_offs": 0.
### Actual behavior
imfile-statefile doesn't reset. and rsyslog stops sending logs.
### Environment
- rsyslog version: 8.2112.0
- platform: Ubuntu 22.04.1 LTS
Rsyslog configuration:

```
$ModLoad imfile
$InputFilePollInterval 10
$PrivDropToGroup adm
$InputFileName /home/user/logs/logfile.out
$InputFileTag logfile
$InputFileStateFile /var/spool/logfille
$InputFileSeverity info
$InputFileFacility local1
$InputRunFileMonitor
$reopenOnTruncate="on"
$InputFilePersistStateInterval 1000
$LocalHostName TEST_RSYSLOG

*.*       @@monitoring.server.com:port
```

logrotate config fime:
```
/home/user/logs/logfile.out
{
  hourly
  size 1K
  compress
  delaycompress
  rotate -1w  dateexte
  dateformat .%Y-%m-%d-%H-%M-%S.gz
  copytruncate
  olddir /home/user/logs/backup
  postrotate
   sudo rm /var/spool/rsyslog/*
   sudo service rsyslog restart
  endscript
}
```
We have a use case where we rotate and track Tomcat webserver logs:
1) copytruncate causes rsyslog to hang
2) nocopytruncate or create causes tomcat to hang (fd still points to the old compressed log)

Solving any of these two should resolve the issue. 

For problem 2 we didn't find any solution and are open for help as we understand the problems with copytruncate.
For problem 1, we tried many things the only thing that worked is "systemctl stop rsyslog; rm /var/spool/rsyslog/*; systemctl start rsyslog"

**Comment by davidelang** [comment]
can you update to a more current version? I think that imfile was re-written 
since Dec 2021

David Lang

**Comment by YMAHPro** [comment]
Hi David, I have updated Rsyslog to version 8.2404.0 restarted Rsyslog the issue still persist, when logrotate is executed Rsyslog hangs (fd files are not reset), and the only command that fixes the issue is "systemctl stop rsyslog; rm /var/spool/rsyslog/*; systemctl start rsyslog"

**Comment by rgerhards** [comment]
Please create a debug log. Instructions: https://www.rsyslog.com/doc/v8-stable/troubleshooting/howtodebug.html

**Be sure that the problem in question occurs at least once during the debug run**. Once this is done, please post the debug log (e.g. via pastebin or github). After that, we can review the contents and check how exactly rsyslog processing worked.

This should include a restart of where the file is not processed.

**Comment by YMAHPro** [comment]
here is the pastebin:
https://pastebin.com/xQsK63d1
this is the behaviour of rsyslog from the server's POV.
![image](https://github.com/rsyslog/rsyslog/assets/171338851/4b4e9864-2588-4df0-ba4b-68ab771e99f7)

PS: I didn't capture executing the command "systemctl stop rsyslog; rm /var/spool/rsyslog/*; systemctl start rsyslog" because I used "$DebugLevel 1" and the command resets the logs

**Comment by YMAHPro** [comment]
the sequence of commands I did: 
systemctl stop rsyslog; rm /var/spool/rsyslog/*; systemctl start rsyslog (wait 15 secs)
kill -USR1 `cat /var/run/rsyslogd.pid` (wait 15 secs)
logrotate -v -f /etc/logrotate.d/catalina ( I capture logs here wait 15 secs)
systemctl stop rsyslog; rm /var/spool/rsyslog/*; systemctl start rsyslog (I make screen capture here).

**Comment by jpalermo** [comment]
We're seeing similar problems too. We're watching multiple files and ones that get logrotated seem to have rsyslog get stuck in the imfile module and no longer forwards messages written to those files.

Verified on the latest version 8.2502.0, but we also have observed the behavior os a few other releases shipped in 2024.

**Comment by davidelang** [comment]
what is your logrotate config? are you sending a HUP to rsyslog after the 
file rotation?

David Lang

**Comment by jpalermo** [comment]
Ahh, interesting. Wasn't aware of that requirement.

It seems like our files that work fine are doing a postrotate of:
```
sudo systemctl kill -s HUP rsyslog.service
```

But the ones that are getting stuck are doing:
```
sudo service rsyslog force-reload
```

I'm guessing `force-reload` either used to work and stopped at some point last year. Or, our users never noticed that some logs were getting stuck until recently and it's never worked correctly.

We'll switch them all to using HUP and see if that fixes the problem.

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant to the original reporter, but I am adding the conclusion for people who arrive here via search while we work through the backlog. We do not close rsyslog issues lightly, and we are not closing this merely because of age.

Triage result: valid non-code answer.

Reasoning: Latest analysis points to logrotate integration: bad cases used `service rsyslog force-reload`, good cases used explicit HUP. Use `systemctl kill -s HUP rsyslog.service` or equivalent and reopen with a current minimal repro if that still fails.

If this still reproduces on a current rsyslog release and maintained platform, please open a new issue with the current version, complete configuration, logs, and a minimal reproducer so we can investigate it as a fresh current problem.
